lms/djangoapps/labster_accounts/management/commands/obfuscate_user_data.py

```diff
@@ -0,0 +1,159 @@
+                    )
+                    not_enrolled_users = CourseEnrollmentAllowed.may_enroll_and_unenrolled(ccx).values_list(
+                        'email', flat=True
+                    )
+
+                    all_users = chain(not_enrolled_users, enrolled_users)
+
+                    user_emails |= set(all_users)
+                except CourseLicense.DoesNotExist:
+                    print("CourseLicense with license code `%s` was not found" % lic)
+
+        if emails:
+            user_emails |= set(emails)
+
+        self.obfuscate_users(user_emails)
+
+    def obfuscate_users(self, user_emails):
+        """
+        Obfuscates user model fields: names, email.
+        """
+        users = User.objects.filter(email__in=user_emails).exclude(email__contains='labster.com')
+
+        field_actions = {
+            'first_name': self.generate_random_string,
+            'last_name': self.generate_random_string,
+            'username': self.generate_random_string,
+            'profile.name': self.generate_random_string,
+            'profile.mailing_address': self.generate_random_string,
+            'email': self.obfuscate_email
+        }
+
+        self.obfuscate_names_emails(users, field_actions)
+
+    @staticmethod
+    def obfuscate_names_emails(query_set, field_actions):
+        """
+        Obfuscates objects name, email and other fields.
+        """
+        fields_select = set([key.split('.')[0] if '.' in key else key for key in field_actions.keys()])
+        for obj in query_set.only(*fields_select):
+            print("Obfuscating data for %s" % obj)
+            values = {}
+            references = {}
+            for field, action in field_actions.items():
+                action = action_decorator(action, field)
+                if '.' in field:
+                    # one2one relation or foreign key found
+                    names = field.split('.')
+                    # get relation field value
+                    val = getattr(getattr(obj, names[0]), names[1], '')
+                    if not (names[0] in references):
+                        references[names[0]] = {}
+                    references[names[0]].update({names[1]: action(val)})
+                else:
+                    val = getattr(obj, field, '')
+                    values[field] = action(val)
+
+            try:
+                with transaction.atomic():
+                    if values:
+                        obj.__class__.objects.filter(id=obj.id).update(**values)
+
+                    if references:
+                        for rel, item in references.items():
+                            _class = getattr(obj, rel).__class__
+                            _id = getattr(obj, rel).id
+                            _class.objects.filter(id=_id).update(**item)
+            except IntegrityError:
+                print("User object with email `%s` can not be obfuscated. Please try one more time" % obj.email)
```

Comment:
Don't we need a stack trace in this case?

Reply:
If we know that the issue with email duplication can we simply re-generate a new email and try one more time?
If 3 tries fail, write an error message and stack trace and continue iwth the next user.